### PR TITLE
fix #1220 Add transactional support with commit/rollback path

### LINKED ISSUE #1220: Explore support for transaction-like operations
This is close to what `using` is doing, expect with the concept of `commit` and `rollback` paths instead of simple resource cleanup.

The original onNext signals should be replayed as is no matter what, but there could be a delay just before the terminal signal due to the commit/rollback operations having to be executed (possibly asynchronously).

An error in the commit publisher would change the terminal signal to an `onError`.
An error in the rollback publisher (triggered by an original error) would still terminated with an `onError`, but exposing the rollback failure first and foremost and suppressing the original rollback cause.

### PR DESCRIPTION
**Pull request by simonbasle**
Need to support fuseable and conditional, and to validate the API.

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/publisher/FluxTransactional.java [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+	final Function<? super S, ? extends Publisher<? extends T>> transactionClosure;
+	final Function<? super S, ? extends Publisher<?>>           commit;
+	final Function<? super S, ? extends Publisher<?>>           rollback;
+	final boolean                                               rollbackOnCancel;
+
+	FluxTransactional(Supplier<S> transactionalResourceSupplier,
+			Function<? super S, ? extends Publisher<?>> commit,
+			Function<? super S, ? extends Publisher<?>> rollback,
+			Function<? super S, ? extends Publisher<? extends T>> transactionClosure,
+			boolean rollbackOnCancel) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxTransactional.java]: replace with a third `Function`, maybe reorder arguments in public API to stick to similar order to using.

**Comment by simonbasle** [comment]
TODO: need `Fuseable` and `Conditional` versions.

**Comment by OlegDokuka** [comment]
I guess there is a need for an extension of transactional that may solve something like following -> https://stackoverflow.com/questions/50686524/how-to-trigger-mono-execution-after-another-mono-terminates/50686674#50686674

**Comment by simonbasle** [comment]
I'm reworking the operator to have a third `Function<RESOURCE, Publisher>` for the `cancel` case, instead of a `boolean` that makes the `error` case a `cancel` case as well.

I think that would cover the use case mentioned by @OlegDokuka, with a bit of overhead since the SO question is asking for a `doFinally` kind of solution: here he'd have to repeat the same function 3 times (complete path, error path and cancel path).

**Comment by simonbasle** [comment]
One open question with the `cancel` case being a separate `Function`: should it default to doing absolutely nothing on `cancel()` if no function is provided? or should it invoke the `complete` function (like I initially did in my PR)?

Keep in mind that eg. `.take(3)` would cancel the main sequence after 3 `onNext`, so in the case of a DB transaction that probably means a `commit()` is needed rather than nothing or a rollback.

**Comment by OlegDokuka** [comment]
@simonbasle Disagree 😇. With current API it would be impossible to implement the mentioned thing. 

For example, embracing the current solution, we will get following 

```java
AtomicInteger counter = new AtomicInteger();
Flux.using(
   () -> Mono.fromCallable(() -> counter.getAndIncrement())),
   (mono) -> mono.then(Mono.just("a")),
   (mono) -> mono.then(),
   (mono) -> mono.then(),
   (mono) -> mono.then(),
)
```

by running that code, instead of the expected counter with value 1, we will get value 2 instead, which means that we cannot supply resource asynchronously and subscriptions to that resource will be done internaly

**Comment by OlegDokuka** [comment]
Regarding the second point, as we have talked about that with my colleague, effectively (in DB transactions) cancelation means rollback. Thus I guess we have to leave an additional hook for cancelation as well (and give in that way provide custom logic here). However, in the case of the absence of that hook, we may call complete signal, since in the case of `take(3)` that will be the success rather than failure. (My opinion)

**Comment by simonbasle** [comment]
> For example, embracing the current solution, we will get following

I don't think that is representative of the SO's original use case (the reactive unlocking is not re-triggering the `stage.get()` `Mono`).

**Comment by OlegDokuka** [comment]
or even more, if we try to do something like following: 

```
Flux.using(
   () -> {
     System.out.println("locked");
     return doLock(params)
   },
   (monoLock) -> monoLock.doOnNext(System.out::println); 	 
                         .then(stage.get().doOnNext(System.out::println)),
   (monoLock) -> monoLock.flatMap(lock -> {
     System.out.println("unlocked" + lock);
     return doUnlock(lock);
   }),
   (monoLock) -> monoLock.flatMap(lock -> doUnlock(lock))
).subscribe();
```

**Comment by simonbasle** [comment]
In his API the only issue is the `doLock` method which seems to perform the lock in an asynchronous fashion. With a little adaptation, his code coud become something more like this (provided async using is also added to `Mono`, which it should):

```java
public interface Locks {

    Mono<ReactiveDistributedLock> doLock(LockParams params);

    Mono<Boolean> doUnlock(ReactiveDistributedLock lock);

    default <T> Mono<T> withLock(LockParams params, Supplier<Mono<T>> stage) {
        return doLock(params) //asynchronously establish the lock?
                .flatMap(lock -> Mono.using(
                    () -> lock, //share lock, hopefully doLock already produces 1 lock per Subscriber
                    lock -> stage.get(), //perform the synchronized action
                    Locks::doUnlock) //unlock on cancel, onError, onComplete
                );
}
```

If the `doLock` can be synchronous, then the whole thing becomes:

```java
public interface Locks {

    ReactiveDistributedLock prepareLock(LockParams params);

    Mono<Boolean> doUnlock(ReactiveDistributedLock lock);

    default <T> Mono<T> withLock(LockParams params, Supplier<Mono<T>> stage) {
        return Mono.using(() -> prepareLock(params), //synchronously establish the lock?
                lock -> stage.get(), //perform the synchronized action
                Locks::doUnlock); //unlock on cancel, onError, onComplete
}
```

**Comment by OlegDokuka** [comment]
right ... 
image

😁

**Comment by OlegDokuka** [comment]
Right.. I guess we may do an alias for that as well (async resource)

**Comment by simonbasle** [comment]
Note: from now on the pushes to this PR should be built and available as snapshots in [Spring's snapshot repository](https://repo.spring.io/snapshot/io/projectreactor/reactor-core/3.2.0.1220.BUILD-SNAPSHOT/) under the version `3.2.0.1220.BUILD-SNAPSHOT`.

**Pull request merged**